---
title: "Lab_10"
author: "Liaqat Ali"
date: "11/8/2020"
output: html_document
---

```{r}
library(vcfR)

## *****       ***   vcfR   ***       *****
##    This is vcfR 1.12.0 
##      browseVignettes('vcfR') # Documentation
##      citation('vcfR') # Citation
##    *****       *****      *****       *****
```


```{r}
data(vcfR_example)
vcf
```

```{r}
strwrap(vcf@meta[1:7])
```

```{r}
queryMETA(vcf)
```

```{r}
queryMETA(vcf, element = 'DP')
```


```{r}
head(getFIX(vcf))
```


```{r}
vcf@gt[1:6, 1:4]
```


```{r}
head(vcf)
```


```{r}
write.vcf(vcf, "myVCFdata_filtered.vcf.gz")
```




```{r}
# 1) How would we find more information about read.vcfR()?
?read.vcfR
```

```{r}
#2) How would we learn what the acronym “AD” stands for?
queryMETA(vcf, element = 'AD')
```

```{r}
#3) We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?

tail(vcf@fix)
```

```{r}
# 4) There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?

plot(vcf)
```


```{r}

library(ggplot2)
qplot(getQUAL(vcf), geom = "histogram")
```

```{r}
# 5) How would we query the sample names?
colnames(vcf@gt)
```




```{r}
library(vcfR)

# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```

```{r}
plot(chrom)
```

```{r}
chromoqc(chrom, dp.alpha = 66)
```

```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
```

```{r}
plot(chrom) 
```


```{r}
chromoqc(chrom, dp.alpha = 66)
```

```{r}
#vcf <- read.vcfR("pinfsc50_qc.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```

```{r}
head(chrom@var.info)
```

```{r}
head(chrom@win.info)
```

```{r}
library(vcfR)
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```

```{r}
knitr::kable(head(myDiff[,16:19]))
```


```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

```{r}
library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)

p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```


```{r}
# Exercises Part II
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))

```

```{r}
table(pop)
```

```{r}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```

```{r}
queryMETA(vcf)
```

